File: Data Structures & Algorithms/partition-equal-subset-sum/submission-1.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        N = len(nums)
        dp = {}

        def dfs(i, l, r):
            if i == N:
                if l == r:
                    dp[(l,r)] = True
                else: 
                    dp[(l,r)] = False
                return dp[(l,r)]
            
            if (l,r) in dp:
                return dp[(l,r)]
            
            dp[(l,r)] = (dfs(i + 1, l + nums[i], r) 
                    or dfs(i + 1, l, r + nums[i]))
            
            return dp[(l,r)]
        
        ans = dfs(0, 0, 0)
        print(dp)
        return ans

File: Data Structures & Algorithms/two-integer-sum/submission-1.java
class Solution {
    public int[] twoSum(int[] nums, int target) {
        int[] result = new int[2];
        Map<Integer, Integer> found = new HashMap<>();
        for(int i = 0; i < nums.length; i++) {
            if(found.containsKey(target-nums[i])) {
                if(found.get(target-nums[i]) > i){
                    result[0] = i;
                    result[1] = found.get(target-nums[i]);
                } else {
                    result[0] = found.get(target-nums[i]);
                    result[1] = i;
                }
                break;
            } 
            found.put(nums[i], i);
        }
        return result;
    }
}
